python/cudf/cudf/core/tools/datetimes.py

```diff
@@ -652,3 +659,281 @@ def _isin_datetimelike(
+        raise NotImplementedError(
+            "Mixing fixed and non-fixed frequency offset is unsupported."
+        )
+    offset = offset.to_pandas()
+
+    # Depending on different combinations of `start`, `end`, `offset`,
+    # `periods`, the following logic makes sure before computing the sequence,
+    # `start`, `periods`, `offset` is defined
+
+    _periods_not_specified = False
+
+    if start is None:
+        end = cudf.Scalar(end, dtype=dtype)
+        start = cudf.Scalar(
+            pd.Timestamp(end.value) - (periods - 1) * offset, dtype=dtype
+        )
+    elif end is None:
+        start = cudf.Scalar(start, dtype=dtype)
+    elif periods is None:
+        # When `periods` is unspecified, its upper bound estimated by
+        # dividing the number of nanoseconds between two timestamps with
+        # the lower bound of `freq` in nanoseconds. While the final result
+        # may contain extra elements that exceeds `end`, they are trimmed
+        # as a post processing step. [1]
+        _periods_not_specified = True
+        start = cudf.Scalar(start, dtype=dtype)
+        end = cudf.Scalar(end, dtype=dtype)
+        _is_increment_sequence = end >= start
+
+        periods = math.ceil(
+            int(end - start) / _offset_to_nanoseconds_lower_bound(offset)
+        )
+
+        if periods < 0:
+            # end < start, return empty column
+            periods = 0
+        elif periods == 0:
+            # end == start, return exactly 1 timestamp (start)
+            periods = 1
+
+    if normalize:
+        old_dtype = start.value.dtype
+        start = cudf.Scalar(
+            start.value.astype("datetime64[D]").astype(old_dtype)
+        )
+
+    # The estimated upper bound of `end` is enforced to compute to make sure
+    # overflow components are thrown before actually computing the sequence.
+    # FIXME: when `end_estim` is out of bound, but the actual `end` is not,
+    # we shouldn't raise but compute the sequence as is. The trailing overflow
+    # part should get trimmed at the end.
+    end_estim = (
+        pd.Timestamp(start.value) + (periods - 1) * offset
+    ).to_datetime64()
+    if "months" in offset.kwds or "years" in offset.kwds:
+        # If `offset` is non-fixed frequency, resort to libcudf.
+        res = libcudf.datetime.date_range(start.device_value, periods, offset)
+    else:
+        # If `offset` is fixed frequency, we treat both timestamps as integers
+        # and evenly divide the given integer range.
+        arr = cp.linspace(
+            start=start.value.astype("int64"),
+            stop=end_estim.astype("int64"),
+            num=periods,
+        )
+        res = cudf.core.column.as_column(arr).astype("datetime64[ns]")
+
+    if _periods_not_specified:
+        # As mentioned in [1], this is a post processing step to trim extra
+        # elements when `periods` is an estimated value.
+        res = res[(res <= end) if _is_increment_sequence else (res <= start)]
```

Comment:
I think this filtration can be avoided for cases where the offset is purely fixed frequency

Reply:
Indeed!